File: app/mobile/native/LocationService.java
package in.recoveriq.app;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Intent;
import android.location.Location;
import android.os.Build;
import android.os.IBinder;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.location.Priority;

import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;

/**
 * Foreground location service. Reads GPS via FusedLocationProvider and POSTs each fix to the
 * server (same /api/tracking/ping endpoint the web app uses). Because it's a foreground service
 * with an ongoing notification, Android keeps it running when the app is backgrounded or locked.
 */
public class LocationService extends Service {
    private static final String TAG = "RQTrack";
    private static final String CHANNEL_ID = "recoveriq_tracking";
    private FusedLocationProviderClient client;
    private LocationCallback callback;
    private String pingUrl;
    private String token;

    @Override
    public void onCreate() {
        super.onCreate();
        createChannel();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null) {
            if (intent.getStringExtra("url") != null) pingUrl = intent.getStringExtra("url");
            if (intent.getStringExtra("token") != null) token = intent.getStringExtra("token");
        }
        Log.i(TAG, "onStartCommand: url=" + pingUrl + " hasToken=" + (token != null && token.length() > 0));
        startForeground(1001, buildNotification());
        startUpdates();
        return START_STICKY;
    }

    private void startUpdates() {
        if (client != null) return;
        client = LocationServices.getFusedLocationProviderClient(this);
        LocationRequest req = new LocationRequest.Builder(Priority.PRIORITY_HIGH_ACCURACY, 5000L)
                .setMinUpdateIntervalMillis(3000L)
                .build();
        callback = new LocationCallback() {
            @Override
            public void onLocationResult(LocationResult result) {
                Location loc = result.getLastLocation();
                if (loc != null) {
                    Log.i(TAG, "location " + loc.getLatitude() + "," + loc.getLongitude() + " acc=" + loc.getAccuracy());
                    postLocation(loc);
                } else {
                    Log.w(TAG, "location result had no location");
                }
            }
        };
        try {
            client.requestLocationUpdates(req, callback, Looper.getMainLooper());
            Log.i(TAG, "requested location updates OK");
        } catch (SecurityException e) {
            Log.e(TAG, "NO LOCATION PERMISSION — cannot start updates", e);
        }
    }

    private void postLocation(final Location loc) {
        if (pingUrl == null || pingUrl.length() == 0) return;
        final String url = pingUrl;
        final String tk = token;
        new Thread(new Runnable() {
            @Override
            public void run() {
                HttpURLConnection c = null;
                try {
                    c = (HttpURLConnection) new URL(url).openConnection();
                    c.setConnectTimeout(15000);
                    c.setReadTimeout(15000);
                    c.setRequestMethod("POST");
                    c.setRequestProperty("Content-Type", "application/json");
                    if (tk != null && tk.length() > 0) c.setRequestProperty("Authorization", "Bearer " + tk);
                    c.setDoOutput(true);
                    String body = "{\"latitude\":" + loc.getLatitude()
                            + ",\"longitude\":" + loc.getLongitude()
                            + ",\"accuracy\":" + loc.getAccuracy()
                            + ",\"speed\":" + loc.getSpeed() + "}";
                    OutputStream os = c.getOutputStream();
                    os.write(body.getBytes("UTF-8"));
                    os.flush();
                    os.close();
                    int code = c.getResponseCode();
                    Log.i(TAG, "POST " + url + " -> HTTP " + code);
                } catch (Exception e) {
                    Log.e(TAG, "POST failed to " + url, e);
                } finally {
                    if (c != null) c.disconnect();
                }
            }
        }).start();
    }

    private Notification buildNotification() {
        return new NotificationCompat.Builder(this, CHANNEL_ID)
                .setContentTitle("RecoverIQ — on duty")
                .setContentText("Sharing your live location with your branch.")
                .setSmallIcon(android.R.drawable.ic_menu_mylocation)
                .setOngoing(true)
                .setPriority(NotificationCompat.PRIORITY_LOW)
                .build();
    }

    private void createChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel ch = new NotificationChannel(
                    CHANNEL_ID, "Location tracking", NotificationManager.IMPORTANCE_LOW);
            NotificationManager mgr = getSystemService(NotificationManager.class);
            if (mgr != null) mgr.createNotificationChannel(ch);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (client != null && callback != null) client.removeLocationUpdates(callback);
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: android-native/app/src/main/java/in/recoveriq/app/location/LocationService.kt
package `in`.recoveriq.app.location

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.pm.ServiceInfo
import android.os.Build
import android.os.IBinder
import android.util.Log
import androidx.core.app.NotificationCompat
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationCallback
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationResult
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.Priority
import `in`.recoveriq.app.MainActivity
import `in`.recoveriq.app.R
import `in`.recoveriq.app.data.Api
import `in`.recoveriq.app.data.PingCreate
import `in`.recoveriq.app.data.Repository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Always-on foreground location service.
 *
 * This is the piece that makes tracking survive the app being backgrounded or the phone
 * locked — a native foreground service with foregroundServiceType=location, unlike the
 * old WebView bridge. It posts each fix to /api/tracking/ping over the shared authed
 * OkHttp client (JWT + cert pinning). START_STICKY + BootReceiver keep it alive.
 */
class LocationService : Service() {

    companion object {
        private const val TAG = "RQTrack"
        const val CHANNEL_ID = "recoveriq_location"
        const val NOTIF_ID = 1001

        const val ACTION_START = "in.recoveriq.app.action.START_TRACKING"
        const val ACTION_STOP = "in.recoveriq.app.action.STOP_TRACKING"

        /** Update cadence. Baseline (e.g. standing still) a fix about every 20s — enough to stay
         *  "live" (presence window is 3 min) without writing a LocationPing row every few seconds
         *  per officer, which hammered the DB's single writer. While MOVING the provider may still
         *  deliver as fast as every 3s, so route tracking stays real-time. */
        private const val INTERVAL_MS = 20_000L
        private const val FASTEST_MS = 3_000L
    }

    private lateinit var client: FusedLocationProviderClient
    private lateinit var repo: Repository
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private val callback = object : LocationCallback() {
        override fun onLocationResult(result: LocationResult) {
            val loc = result.lastLocation ?: run {
                Log.w(TAG, "location result had no location")
                return
            }
            Log.i(TAG, "fix ${loc.latitude},${loc.longitude} acc=${loc.accuracy}")
            postLocation(loc.latitude, loc.longitude, loc.accuracy.toDouble(),
                if (loc.hasSpeed()) loc.speed.toDouble() else null)
        }
    }

    override fun onCreate() {
        super.onCreate()
        client = LocationServices.getFusedLocationProviderClient(this)
        repo = Repository(applicationContext)
        createChannel()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent?.action == ACTION_STOP) {
            Log.i(TAG, "onStartCommand: STOP")
            stopSelf()
            return START_NOT_STICKY
        }

        Log.i(TAG, "onStartCommand: START")
        startAsForeground()

        // Make sure the shared token cache is hydrated even after a cold process restart.
        scope.launch { repo.bootstrapToken() }
        requestUpdates()
        return START_STICKY
    }

    private fun startAsForeground() {
        val notif = buildNotification()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            startForeground(NOTIF_ID, notif, ServiceInfo.FOREGROUND_SERVICE_TYPE_LOCATION)
        } else {
            startForeground(NOTIF_ID, notif)
        }
    }

    private fun requestUpdates() {
        val request = LocationRequest.Builder(Priority.PRIORITY_HIGH_ACCURACY, INTERVAL_MS)
            .setMinUpdateIntervalMillis(FASTEST_MS)
            .setWaitForAccurateLocation(false)
            .build()
        try {
            client.requestLocationUpdates(request, callback, mainLooper)
            Log.i(TAG, "requested location updates OK")
        } catch (e: SecurityException) {
            Log.e(TAG, "NO LOCATION PERMISSION — cannot start updates", e)
            stopSelf()
        }
    }

    private fun postLocation(lat: Double, lng: Double, acc: Double?, speed: Double?) {
        scope.launch {
            try {
                if (Api.token.isNullOrBlank()) repo.bootstrapToken()
                if (Api.token.isNullOrBlank()) {
                    Log.w(TAG, "no token yet — skipping POST")
                    return@launch
                }
                Api.service.ping(PingCreate(latitude = lat, longitude = lng, accuracy = acc, speed = speed))
                Log.i(TAG, "POST ping OK")
            } catch (e: Exception) {
                Log.e(TAG, "POST ping failed: ${e.message}", e)
            }
        }
    }

    private fun buildNotification(): Notification {
        val open = PendingIntent.getActivity(
            this, 0,
            Intent(this, MainActivity::class.java),
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
        )
        return NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle(getString(R.string.tracking_notification_title))
            .setContentText(getString(R.string.tracking_notification_text))
            .setSmallIcon(R.drawable.ic_stat_location)
            .setOngoing(true)
            .setContentIntent(open)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .build()
    }

    private fun createChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val ch = NotificationChannel(
                CHANNEL_ID,
                getString(R.string.tracking_channel_name),
                NotificationManager.IMPORTANCE_LOW
            )
            ch.setShowBadge(false)
            getSystemService(NotificationManager::class.java).createNotificationChannel(ch)
        }
    }

    override fun onDestroy() {
        Log.i(TAG, "onDestroy — removing updates")
        client.removeLocationUpdates(callback)
        scope.cancel()
        super.onDestroy()
    }

    override fun onBind(intent: Intent?): IBinder? = null
}
